### PR TITLE
[Enhancement] Support rewrite groupingsets to union all

### LINKED ISSUE #10480: Optimize rollup/groupingsets to union all
## Enhancement
we can optimize rollup/groupingsets to union all to avoid large memory use for hash aggregate.

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [ ] bugfix
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #10480 

I had tested this pr in TPCDS 100GB with 3*BE(64C), results are below:
```
		dop=0		dop=1		dop=4		dop=8	
		off 	on	off 	on	off 	on	off 	on
query05.sql	1.077	1.117	1.407	1.333	0.563	0.58	0.518	0.555
query14.sql	8.482	8.733	18.076	18.306	5.944	6.005	4.708	4.727
query18.sql	0.865	1.266	1.262	0.966	0.461	0.395	0.344	0.355
query22.sql	1.922	2.041	22.409	4.599	6.7	1.57	3.963	1.567
query27.sql	0.418	0.547	1.152	1.14	0.386	0.39	0.266	0.289
query36.sql	0.396	0.526	0.906	0.842	0.31	0.288	0.227	0.219
query67.sql	6.647	5.951	77.813	22.163	23.753	6.451	14.063	4.618
query70.sql	0.726	1.139	6.471	3.501	1.821	1.08	1.08	0.965
query77.sql	0.736	0.81	0.891	0.908	0.334	0.345	0.281	0.314
query80.sql	1.085	1.183	2.59	2.574	0.97	0.946	0.717	0.757
query86.sql	0.408	0.609	1.448	0.489	0.436	0.195	0.282	0.263
		22.762	23.922	134.425	56.821	41.678	18.245	26.449	14.629	
```

When pipeline_dop is limited(eg dop=8), the GroupingSets Rewrite will get a better result than the current RepeatNode implement.

However I had noticed that when using CTE+UnionAll, the framework does not work well for each dop which is another problem. I will think about it later.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 5 / 89 (05.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|0|83|00.00%|[52, 53, 54, 55, 59, 60, 65, 66, 68, 69, 70, 72, 73, 74, 75, 79, 81, 82, 83, 84, 85, 88, 89, 90, 93, 94, 95, 96, 97, 98, 100, 101, 102, 103, 107, 108, 109, 111, 112, 113, 114, 115, 117, 118, 119, 120, 122, 124, 125, 126, 127, 128, 131, 132, 135, 138, 139, 140, 141, 142, 143, 144, 145, 147, 149, 150, 155, 156, 159, 160, 161, 162, 163, 164, 167, 168, 169, 170, 171, 172, 173, 174, 177]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|3|66.67%|[240]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by LiShuMing** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Comment by kangkaisen** [comment]
need performance report

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 1]
```diff
@@ -536,6 +537,9 @@ public class SessionVariable implements Serializable, Writable, Cloneable {
     @VariableMgr.VarAttr(name = CBO_ENABLE_LOW_CARDINALITY_OPTIMIZE)
     private boolean enableLowCardinalityOptimize = true;
 
+    @VariableMgr.VarAttr(name = ENABLE_OPTIMIZER_REWRITE_GROUPINGSETS_TO_UNION_ALL)
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:540]: set invisible

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+ *      select null, b, count(1) from t group by b
+ *      union all
+ *      select null, null, count(1) from t
+ */
+public class RewriteGroupingSetsByCTERule extends TransformationRule {
+
+    public RewriteGroupingSetsByCTERule() {
+        super(RuleType.TF_REWRITE_GROUPING_SET,
+                Pattern.create(OperatorType.LOGICAL_AGGR)
+                        .addChildren(Pattern.create(OperatorType.LOGICAL_REPEAT, OperatorType.LOGICAL_PROJECT)));
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java:54]:
```suggestion
                        .addChildren(Pattern.create(OperatorType.LOGICAL_REPEAT, OperatorType.PATTERN_LEAF)));
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+ * Rewrite grouping sets to union all. eg:
+ *      select a, b, count(1) from t group by rollup(a, b);
+ * rewrite to:
+ *      select a, b, count(1) from t group by a, b
+ *      union all
+ *      select a, null, count(1) from t group by a
+ *      union all
+ *      select null, b, count(1) from t group by b
+ *      union all
+ *      select null, null, count(1) from t
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java:47]: can we rewrite to ?
```

     with xx (select a, b, count(1) cc from t group by a, b)
     select a, b, cc from xx;
     union all
     select a, null, sum(cc) from xx group by a
     union all
     select null, b, sum(cc) from xx group by b
     union all
     select null, null, sum(cc) from xx
```

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java:47]: Yep. The document is wrong, I'll fix it.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 85 / 92 (92.39%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteGroupingSetsByCTERule.java|77|84|91.67%|[168, 169, 170, 171, 172, 173, 174]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|

**Pull request merged**